### DISCUSSION TITLE
New things in the editor

### DISCUSSION THREAD

**Discussion by Luffyyy** [diesel-modding/BeardLib-Editor#636] [General]
## 4.4
In 4.4 the way levels work has changed, instead of loading from the <include> node, BeardLib automatically reads ModPath/levels/LevelName/add_local.xml

### Why?
* In order to decrease the size of the main.xml. It can get quite large.
* Include was sorta adding files while not using the AddFiles syntax, making it a little confusing.
* Joining the 1st point, trying to add things like cubemaps to it... yuck.

### What do I need to do?
Mostly...nothing. The editor will ask you to convert your map to the new format once you try to save it. This will force min_lib_ver to be 4.5 (if it doesn't exist or it's lower).

## 4.5
Instances were actually broken with the new format, but luckily no map has used it from what it seems. This was fixed with this version. 
In 4.5 levels and instances got another change - an internal one. The internal path will be as follows: levels/mods/ModName/LevelName and levels/instances/mods/ModName/InstanceName this takes care of name conflicts mainly for **instances**.

EditableGUIs will now have a space added to them. This fixes this issue: #557. This will be later handled by an added parameter which will remove the additional space. The reason for this is the way custom XML is handled in Diesel and it has no way of knowing whether 00 or 001 is a number or a string. Sadly, it defaults to a number, which is less fixable if it were to assume numbers as strings (we could just do tonumber then!) Anyhow, we trick it by adding a space to think it's a string, and it worked. It's a little weird, but what do you expect from Diesel :)

Massunits (brushes) support was added - probably one of the most requested features to the editor and was held due to thinking it would be too hard to port. I was proven wrong (easier than the mess that effects editor is). That being said, it's thanks to some work I did few years ago on figuring out the massunits format that allowed me to do this. Also figuring out how to correctly pull out quaternions out of the game was a challenge, I found out we could just read custom XML and that'd allow me to get the information I needed.

### What do I need to do?
At the moment if you use min library version of 4.5, BeardLib will assume you want to use this naming scheme. You will need to change each continent using a custom instance to that format. It's possible the editor will assist in doing so, we'll see.

## Developer Branch
SBLT got a **very** important feature added to it in its last major update - DB loading. It has been a huge pain to try and load anything from the database and we had to think of ways to load things. The editor used to load things from "extract", basically requiring the user to wait and extract 60GB worth of assets just to use the editor, outrageous! 

This has changed and allowed me to rethink and simplify many parts of the editor. Spawning will be smoother!

* New icons - Courtesy of Soosh. The editor feels a little refreshed with new icons that look great.
* More customization - I've added some additional options for customizing the look and feel of the editor.
* Quick actions - Added by @Xeletron - Access options that are often used quicker. 
![image](https://user-images.githubusercontent.com/11510099/126648102-73b288c8-5a56-4b9f-869c-5f0e1296045a.png)
* Physics simulation for untis by @connieprice - Put units in a more natural position more easily!
* New spawning and selecting menu - inspired by an old iteration of the editor ([spoiler](https://cdn.discordapp.com/attachments/633631849633153045/858071893357494282/BDFD17A494B726BAFFCD8167334B6D3C89668DCF.png)). This is visiting an old idea and incorporating it a lot better. The list will update and search quickly compared to the older one, also fitting its items better. 
![image](https://user-images.githubusercontent.com/11510099/126648287-9642560a-a572-4323-8c7f-9110e642e57e.png)
* Massunits editor - like I said massunits support was added and with it support for it in the editor. The editor will allow you to easily paint them over walls, floors and such without much hassle. 
![image](https://user-images.githubusercontent.com/11510099/126648346-9bb9f3c6-a139-4253-b328-a165d8af5e20.png)
* Faster 'World' menu. The world menu's tabs will no longer slow you down. The editor will load each page in the background and avoid destroying it to increase performance. 
* New tips section. You can add new ones by submitting a pull request to Assets/tips.txt
![image](https://user-images.githubusercontent.com/11510099/126648646-2a30f83a-fbfb-4f4b-997d-593a2c1bf481.png)
* Cleaner options menu & moving some less often used options out of the in-editor options menu
![image](https://user-images.githubusercontent.com/11510099/126648416-764d45cc-6954-435a-90cf-f407aeb20c5f.png) 

And many more things that I couldn't bother listing like bug fixes. The update should soon arrive as I'm writing this.

## What's next?
Dunno. I've been focusing on the editor for some weeks, but my time isn't unlimited and I wish to do other work like MWS V3 and studying in general. I'll probably be fixing more bugs than implementing features again, but anyone is welcome to contribute and improve the editor!

**Comment by Luffyyy** [comment]
Might do more stuff like that, but point being is giving more info about changes that might affect map makers and give them a heads up on some things. I might've missed some things (I'm a human after all 😭) I'll add if anything.

**Reply by MiamiCenterPL** [reply]
Mention MassUnits too ;)
